OutOfSchool/OutOfSchool.WebApi/Services/Elasticsearch/ElasticsearchSynchronizationService.cs

```diff
@@ -0,0 +1,183 @@
+{
+    /// <summary>
+    /// Implements the operations for synchronization databases.
+    /// </summary>
+    public class ElasticsearchSynchronizationService : IElasticsearchSynchronizationService
+    {
+        // TODO: move to config
+        private const int NumberOfOperation = 2;
+
+        private readonly IWorkshopService databaseService;
+        private readonly IElasticsearchSyncRecordRepository elasticsearchSyncRecordRepository;
+        private readonly IElasticsearchProvider<WorkshopES, WorkshopFilterES> esProvider;
+        private readonly ILogger<ElasticsearchSynchronizationService> logger;
+        private readonly IMapper mapper;
+
+        public ElasticsearchSynchronizationService(
+            IWorkshopService workshopService,
+            IElasticsearchSyncRecordRepository elasticsearchSyncRecordRepository,
+            IElasticsearchProvider<WorkshopES, WorkshopFilterES> esProvider,
+            ILogger<ElasticsearchSynchronizationService> logger,
+            IMapper mapper)
+        {
+            this.databaseService = workshopService;
+            this.elasticsearchSyncRecordRepository = elasticsearchSyncRecordRepository;
+            this.esProvider = esProvider;
+            this.logger = logger;
+            this.mapper = mapper;
+        }
+
+        public async Task<bool> Synchronize()
+        {
+            logger.LogInformation($"Synchronization of elasticsearch has started.");
+
+            var elasticsearchSyncRecords = await elasticsearchSyncRecordRepository.GetByEntity(
+                ElasticsearchSyncEntity.Workshop,
+                NumberOfOperation).ConfigureAwait(false);
+
+            var resultCreate = await SynchronizeAndDeleteRecords(elasticsearchSyncRecords, ElasticsearchSyncOperation.Create).ConfigureAwait(false);
+            if (!resultCreate)
+            {
+                logger.LogError($"Syncronization of Elasticsearch has failed during {ElasticsearchSyncOperation.Create} operation");
+                return false;
+            }
+
+            var resultUpdate = await SynchronizeAndDeleteRecords(elasticsearchSyncRecords, ElasticsearchSyncOperation.Update).ConfigureAwait(false);
+            if (!resultUpdate)
+            {
+                logger.LogError($"Syncronization of Elasticsearch has failed during {ElasticsearchSyncOperation.Update} operation");
+                return false;
+            }
+
+            var resultDelete = await SynchronizeAndDeleteRecords(elasticsearchSyncRecords, ElasticsearchSyncOperation.Delete).ConfigureAwait(false);
+            if (!resultDelete)
+            {
+                logger.LogError($"Syncronization of Elasticsearch has failed during {ElasticsearchSyncOperation.Delete} operation");
+                return false;
+            }
+
+            logger.LogInformation($"Synchronization of Elasticsearch has finished successfully.");
+
+            return true;
+        }
+
+        public async Task Create(ElasticsearchSyncRecordDto dto)
```

Comment:
do we really need a DTO? we're using this method in the same class but we spend processor time converting dto-> entity as well as instantiate 2 classes (dto and entity).